src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs

```diff
@@ -5,19 +5,193 @@
     {
-        public bool IsSupported(Microsoft.CodeAnalysis.Workspace workspace)
+        private readonly IServiceProvider _serviceProvider;
+
+        private readonly object _gate = new object();
+        private SolutionId _currentSolutionId = null;
+        private string _currentWorkingFolderPath = null;
+        private bool _solutionEventsAdvised = false;
+
+        public event EventHandler<PersistentStorageLocationChangingEventArgs> StorageLocationChanging;
+
+        [ImportingConstructor]
+        public VisualStudioPersistentStorageLocationService([Import] SVsServiceProvider serviceProvider)
+            : base(assertIsForeground: false)
+        {
+            _serviceProvider = serviceProvider;
+        }
+
+        public bool IsSupported(Workspace workspace)
             => workspace is VisualStudioWorkspaceImpl;
 
-        public string GetStorageLocation(Solution solution)
+        public string TryGetStorageLocation(SolutionId solutionId)
+        {
+            lock (_gate)
+            {
+                if (solutionId == _currentSolutionId)
+                {
+                    return _currentWorkingFolderPath;
+                }
+            }
+
+            return null;
+        }
+
+        internal void UpdateForVisualStudioWorkspace(Workspace visualStudioWorkspace)
+        {
+            AssertIsForeground();
+
+            lock (_gate)
+            {
+                if (visualStudioWorkspace.CurrentSolution.Id == _currentSolutionId && _currentWorkingFolderPath != null)
+                {
+                    return;
+                }
+
+                var solution = _serviceProvider.GetService<SVsSolution, IVsSolution>();
+
+                if (!_solutionEventsAdvised)
+                {
+                    solution.AdviseSolutionEvents(new EventSink(this, visualStudioWorkspace), out uint cookie);
+                    _solutionEventsAdvised = true;
+                }
+
+                try
+                {
+                    var solutionWorkingFolder = (IVsSolutionWorkingFolders)solution;
+                    solutionWorkingFolder.GetFolder(
+                        (uint)__SolutionWorkingFolder.SlnWF_StatePersistence, Guid.Empty, fVersionSpecific: true, fEnsureCreated: true,
+                        pfIsTemporary: out var temporary, pszBstrFullPath: out var workingFolderPath);
+
+                    if (!temporary && !string.IsNullOrWhiteSpace(workingFolderPath))
+                    {
+                        OnWorkingFolderChanging_NoLock(
+                            new PersistentStorageLocationChangingEventArgs(
+                                visualStudioWorkspace.CurrentSolution.Id,
+                                workingFolderPath,
+                                mustUseNewStorageLocationImmediately: false));
+                    }
+                }
+                catch
```

Comment:
NFW?

Reply:
@heejaechang It was previously this way. Thoughts?

Reply:
at the time this was written, we didn't have NFW. now we have, so I think it is okay to capture NFW? 

NFW will only be an UI block issue when someone actually asking for full dump, otherwise, it is most likely not noticeable since all it does is sending out some telemetry.